### PR TITLE
Add support for three and four string argument versions of System.String.Concat

### LINKED ISSUE #23859: Entity Framework Core 5.0 does not support string.Concat with more than 2 args
_This issue has been moved from [a ticket on Developer Community](https://developercommunity2.visualstudio.com/t/Entity-Framework-Core-50-does-not-suppo/1285537)._

---
<p>A Query similar to:</p><pre>From t In Me.DataStore.qryAbo
Where t.AboAdrKey = 1035
Order By t.AboActive Descending, t.PrdName, t.TypName
Select New With {t.AboKey, t.AboAdrKey, t.AboActive, . PrdName = String.Concat(t.PrdName, &quot; &quot;, t.TypName, If(t.AboTrial, &quot; (Trial)&quot;, &quot;&quot;))}
</pre><p>gets translated to</p><pre>DECLARE @__DataRow_AdrKey_0 int = 1035;
SELECT [q]. [AboKey], [q]. [AboAdrKey], [q]. [AboActive], [q]. [PrdName], [q]. [TypName], CASE WHEN COALESCE([q].[ AboTrial], CAST(0 AS bit)) = CAST(1 AS bit) THEN N' (Trial)'
    ELSE N''
END
FROM [qryAbo] AS [q]
WHERE COALESCE(CASE
    WHEN ([q].[ AboAdrKey] = @__DataRow_AdrKey_0) AND [q]. [AboAdrKey] IS NOT NULL THEN CAST(1 AS bit)
    ELSE CAST(0 AS bit)
END, CAST(0 AS bit)) = CAST(1 AS bit)
ORDER BY [q]. [AboActive] DESC, [q]. [PrdName], [q]. [TypName]
</pre><p>which brings back 6 columns instead of 4.</p> <p>Is this by design or should System.Concat be supported?</p> <p>Thank you for feedback.</p>

---
#### May Lei [MSFT] on 12/13/2020, 11:39 PM: 

<p>Not sure the area path, please help to reassign if the area path is not correct. Thanks.</p>

#### Travico Support on 12/15/2020, 05:19 AM: 

(private comment, text removed)

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by wmeints**
This PR fixes #23859 by adding support for three and four string argument versions of [String.Concat](https://docs.microsoft.com/en-us/dotnet/api/system.string.concat?view=net-5.0).

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/efcore?pullRequest=23873) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/efcore?pullRequest=23873) before we can accept your contribution.<br/><br/>:x: wmeints [sign now](https://cla.dotnetfoundation.org/dotnet/efcore?pullRequest=23873)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/efcore?pullRequest=23873) it.</sub>

File: test/EFCore.Relational.Tests/Query/Internal/StringMethodTranslatorTest.cs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+using System.Collections.Generic;
+using Microsoft.EntityFrameworkCore.Diagnostics;
+using Microsoft.EntityFrameworkCore.Query.SqlExpressions;
+using Microsoft.EntityFrameworkCore.Storage;
+using Microsoft.EntityFrameworkCore.TestUtilities;
+using Xunit;
+
+namespace Microsoft.EntityFrameworkCore.Query.Internal
+{
+    public class StringMethodTranslatorTest
```
Comment (smitpatel) [inline test/EFCore.Relational.Tests/Query/Internal/StringMethodTranslatorTest.cs:11]: Please remove this test and add test here
https://github.com/dotnet/efcore/blob/dab50a76db2d495126f418363b593cdf01c9c60a/test/EFCore.Specification.Tests/Query/NorthwindWhereQueryTestBase.cs#L1612-L1621
You should be able to replicate that test to have use a string.Concat with more arguments.

**Comment by wmeints** [comment]
Hi, I noticed that this PR hasn't seen any attention from the team. Do you need anything else from me?
I'm happy to help out :-)

**Review by smitpatel — APPROVED** [review]

**Comment by smitpatel** [comment]
@wmeints - Thank you for contribution.

**Pull request merged**